automation, CI: Introduce basic e2e test infrastructure

EdDev opened the pull request:
A dedicated e2e test target has been added to the `make.sh` script.
It delegates all arguments to the new `e2e.sh` script which assists in
deploying a cluster (using `kind`) and `kiagnose`.

The new e2e script also includes a basic sanity validation that deploys
and runs the `echo` checkup.

To run the full set of e2e targets, run:
```
./automation/make.sh --e2e -- \
  --install-kind \
  --install-kubectl \
  --create-cluster \
  --deploy-kiagnose \
  --run-tests \
  --delete-cluster`
```

This PR also includes running the same on CI.

Notes:
- This is a minimalist e2e test validation which should be used as a basic sanity check, assuring the success path for kiagnose.
It is a temporary step until an e2e test suite is introduced.
- The current (explicit) polling could have been improved by using `kubectl wait` command, however I was unable to make it work. Contributing a solution to this is welcome.
- Running the checkup job seems to take a lot of time. We should find the root cause of this and try to reduce it to a minimum (is it the image size?).

orelmisan commented on automation/e2e.sh:
```diff
@@ -0,0 +1,140 @@
+ARGCOUNT=$#
+
+CRI=${CRI:-podman}
+KUBECTL_PATH=${KUBECTL_PATH:-/usr/bin/kubectl}
```
Please consider renaming this `KUBECTL_BINARY` or something along these lines.

orelmisan replied:
Yes.

orelmisan commented on automation/e2e.sh:
```diff
@@ -0,0 +1,140 @@
+
+if [ -n "${OPT_INSTALL_KIND}" ]; then
+    KIND_VERSION=v0.12.0
+    KIND_PATH=${KIND_PATH:-/usr/bin/kind}
```
Please consider renaming this `KIND_BINARY` or something along these lines.

orelmisan commented on automation/e2e.sh:
```diff
@@ -0,0 +1,140 @@
+    --delete-cluster)
+        OPT_DELETE_CLUSTER=1
+        ;;
+    --deploy-kiagnose)
```
Please consider adding an option to remove kiagnose (without destroying the cluster).
Although it seems redundant because there is a step to destroy to whole cluster, IMO this functionality should be checked as well.

orelmisan replied:
It can wait to a follow up PR.

orelmisan reviewed (commented):
Thank you for the PR @EdDev!
It will be very beneficial for the project to have.

Please see the inline comments.

EdDev commented:
[change](https://github.com/kiagnose/kiagnose/compare/5a916f939bff58caafabfe0975b3a5d613b67d5c..e26d6c37cc256b4cf8fda756a9880ed1e7d2526a): Answered the review comments regarding the `kubectl*` and `kind*` variables locations and consistent usage.

orelmisan commented on automation/e2e.sh:
```diff
@@ -0,0 +1,140 @@
+fi
+
+if [ -n "${OPT_DEPLOY_KIAGNOSE}" ]; then
+    ${KUBECTL_PATH} apply -f manifests/kiagnose.yaml
```
I agree that the script shouldn't change the `PATH`.
If you use the full path all over, the `PATH` is irrelevant.
Fine with me.

orelmisan commented on automation/e2e.sh:
```diff
@@ -0,0 +1,140 @@
+        curl -LO https://dl.k8s.io/release/${KUBECTL_VERSION}/bin/linux/amd64/kubectl && \
+        chmod +x ./kubectl && \
+        mv ./kubectl "${KUBECTL_PATH}"
+        echo "kubectl installed successfully at ${KUBECTL_PATH}"
```
It was a copy-pate mistake, sorry.

If you intend to use the full path to kubectl, this comment is irrelevant.

EdDev commented:
change: Rename the variables names and quiet the assertion output of the checkup result (adding `-q` to the `grep` command).

orelmisan commented on automation/e2e.sh:
```diff
@@ -0,0 +1,135 @@
+ARGCOUNT=$#
+
+KUBECTL_VERSION=${KUBECTL_VERSION:-v1.23.0}
+KUBECTL=${KUBECTL:-/usr/bin/kubectl}
```
This path requires a root privileges to install on the first time.
Can you please describe it in the commit message and PR description?

Furthermore I suggest to move it to `/usr/local/bin` or `$HOME/bin` (and then root privileges are not needed, just changing the `PATH` which should be done by the user prior to running the script).

orelmisan commented on automation/e2e.sh:
```diff
@@ -0,0 +1,135 @@
+KUBECTL=${KUBECTL:-/usr/bin/kubectl}
+
+KIND_VERSION=${KIND_VERSION:-v0.12.0}
+KIND=${KIND:-/usr/bin/kind}
```
Same.

orelmisan commented on automation/e2e.sh:
```diff
@@ -0,0 +1,135 @@
+fi
+
+if [ -n "${OPT_CREATE_CLUSTER}" ]; then
+    if ! ${KIND} get clusters | grep kind; then
```
Please consider refactoring the cluster name to a variable.

orelmisan commented on automation/e2e.sh:
```diff
@@ -0,0 +1,135 @@
+
+    echo "Post echo checkup ConfigMap & Job:"
+    echo
+    ${KUBECTL} create -f manifests/echo-checkup.yaml
```
The initial (line 112) and final (line 130) `cd` commands could be dropped if this line changes to:
```bash
${KUBECTL} apply -f ./checkups/echo/manifests/echo-checkup.yaml
```

orelmisan commented on automation/e2e.sh:
```diff
@@ -0,0 +1,135 @@
+    echo
+    ${KUBECTL} create -f manifests/echo-checkup.yaml
+
+    ${KUBECTL} wait --for=condition=complete --timeout=20s job.batch/echo-checkup1 -n kiagnose
```
This command doesn't catch the case in which the framework fails.

orelmisan replied:
Can you please refactor the namespace and Job names into two variables?

orelmisan reviewed (commented):
Thanks for the changes.
Please see the inline comments.

EdDev commented:
change: Added info to the commit message, changed the default binaries installation to the place where the script was called from and used variables for some values. All per the reviewer request.

orelmisan reviewed (approved):
Thank you!
